src/views/ApartmentApp/components/StudentApplication/index.js

```diff
@@ -0,0 +1,794 @@
+ * @typedef { import('../../../../services/housing').ApplicationDetails } ApplicationDetails
+ */
+
+/**
+ * Renders the page for the student apartment application
+ * @param {Object} props The React component props
+ * @param {*} props.authentication The user authentication
+ * @param {StudentProfileInfo} props.userProfile The student profile info of the current user
+ * @returns {JSX.Element} JSX Element for the student application web page
+ */
+const StudentApplication = ({ userProfile, authentication }) => {
+  const [loading, setLoading] = useState(true);
+  const [saving, setSaving] = useState(false);
+  const [unsavedChanges, setUnsavedChanges] = useState(false);
+
+  /** @type {[ApplicationDetails, React.Dispatch<React.SetStateAction<ApplicationDetails>>]} */
+  const [applicationDetails, setApplicationDetails] = useState({
+    ApplicationID: null,
+    DateSubmitted: null,
+    DateModified: null,
+    EditorUsername: null,
+    EditorEmail: null,
+    Applicants: [],
+    ApartmentChoices: [],
+  });
+  /** @type {[Number, React.Dispatch<React.SetStateAction<Number>>]} */
+  const [applicationID, setApplicationID] = useState(null); // Default value of -1 indicate to backend that the application ID number is not yet known
+  const [dateSubmitted, setDateSubmitted] = useState(null); // The date the application was submitted, or null if not yet submitted
+  const [dateModified, setDateModified] = useState(null); // The date the application was last modified, or null if not yet saved/modified
+  const [editorUsername, setEditorUsername] = useState(null); // The username of the application editor
+  /** @type {[ApartmentApplicant[], React.Dispatch<React.SetStateAction<ApartmentApplicant[]>>]} Array of applicant info */
+  const [applicants, setApplicants] = useState([]);
+  /** @type {[ApartmentChoice[], React.Dispatch<React.SetStateAction<ApartmentChoice[]>>]} Array of apartment choice info */
+  const [preferredHalls, setPreferredHalls] = useState([]); // Properties 'HallName' and 'HallRank' must be capitalized to match the backend
+
+  const [agreements, setAgreements] = useState(false); // Represents the state of the agreements card. True if all checkboxes checked, false otherwise
+
+  const [applicationCardsOpen, setApplicationCardsOpen] = useState(false);
+  const [newEditorProfile, setNewEditorProfile] = useState(null); // Stores the StudentProfileInfo of the new editor before the user confirms the change
+  const [changeEditorDialogOpen, setChangeEditorDialogOpen] = useState(false);
+  const [submitDialogOpen, setSubmitDialogOpen] = useState(false); // Use this for submitting app (later feature)
+  const [snackbarOpen, setSnackbarOpen] = useState(false);
+  const [snackbarText, setSnackbarText] = useState('');
+  const [snackbarSeverity, setSnackbarSeverity] = useState('info');
+  const [saveButtonAlertTimeout, setSaveButtonAlertTimeout] = useState(null);
+
+  /**
+   * Loads the user's saved apartment application, if one exists
```

Comment:
This only needs to run when the component first loads (and if the logged in user changes), correct?

The way it's currently running is causing all sorts of issues with your local state when you try to load a saved application.

Reply:
Yes, that is the intended behavior (run at first load). I have been struggling to actually get that you work, and I have been having problems with the `setLoading(false)` being run before the things actually finish loading, giving the user one of the error views from `src/views/ApartmentApp/index.js` rather than the GordonLoader while the page is loading.

I think this problem should first be addressed in `src/views/ApartmentApp/index.js` because some of the problems starting there are causing the weirdness in StudentApplication, but I agree both need to be fixed

Any help/suggestions would be much appreciated, and also I suggest looking at #1063 in case I tried anything in there that relates to this problem